Sources/Testing/Running/Runner.Context.swift

```diff
@@ -0,0 +1,204 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+extension Runner {
+  /// A type which collects the task-scoped context for a running ``Runner``
+  /// instance, the tests it runs, and other objects it interacts with.
+  ///
+  /// This type is intended for use via the task-local
+  /// ``Runner/_Context/current`` property.
+  fileprivate struct _Context: Sendable {
+    /// The runner that is running on the current task, if any.
+    var runner: Runner?
+
+    /// The test that is running on the current task, if any.
+    var test: Test?
```

Comment:
Sharing all these values in a single task-local is going to change the behaviour of tests/runners nested in other runners (i.e. our own tests.) It means that when we temporarily swap out the configuration to run our own unit tests, we'll be swapping out more state than just the configuration, and we need to make sure that a test failure in our own test target can handle that correctly like before.

Reply:
I confirmed (via a new unit test I added) that the handling of this is correct, and in fact I think the new approach is _more_ correct than before: Previously we only swapped out the current `Configuration`, but now, because this group of values is stored together as one `Runner.Context` instance under a single `@TaskLocal`, I can easily swap the entire set of them to their previous value. 

To illustrate: Before this PR, the following test would've failed:

```swift
  @Test func contextScopedEventHandler() async {
    var configuration = Configuration()
    configuration.eventHandler = { _, _ in
      #expect(Test.current?.name == "contextScopedEventHandler()")  // ❌ got "foo" instead
    }

    await Test(name: "foo") {}.run(configuration: configuration)
  }

```

Reply:
Okay, I'm satisfied with that explanation!